linera-execution/src/lib.rs

```diff
@@ -731,6 +731,30 @@ pub trait ContractRuntime: BaseRuntime {
     /// Adds a new item to an event stream. Returns the new event's index in the stream.
     fn emit(&mut self, name: StreamName, value: Vec<u8>) -> Result<u32, ExecutionError>;
 
+    /// Reads an event from a stream. Returns the event's value.
```

Comment:
Does absence become an error? If so, please add that to a comment.

I'd personally prefer if it was an `Option<Vec<u8>>` but we can leave that for a separate PR.

Reply:
It has to be an error, because we want to allow reading events in fast blocks eventually. For that, we'll need to make execution fully deterministic, i.e. the outcome cannot depend on whether a validator is already seeing the event or not. Instead, it needs to return an `EventNotFound` error so that the client can re-send the event (i.e. the block that created it) and retry.

Reply:
I don't understand how returning a `None` - instead of an `Error` - affects the (non)determinism but I can imagine that how the return value is _handled_ could affect it

Reply:
The runtime method doesn't return an error. It fails the block execution.